Allow tmpfiles to create files in a root under an unprivileged directory

dm0- opened the pull request:
I hit a problem with building packages that run tmpfiles to generate the initial package contents in Gentoo, where parent directories are created but not the files themselves.  The issue started for copying files at 16ba55adb1d60577b321c781d81a410da3d28dcd, creating directories at 4c39d899ff00e90b7290e4985696f321d7f2726f, etc.  It turns out to be due to symlink safety checks that prevent running in a root-owned directory under a directory owned by an unprivileged user, which happens to be the case here:

- `/var/tmp/portage/$PACKAGE` = user-owned working directory for packaging
- `/var/tmp/portage/$PACKAGE/image` = root-owned staging directory with final installed file permissions

This also happens if you try to set up a new disk image manually in a non-root-owned directory.  Here is a minimal reproducer of this:

    truncate --size=1G hda.img
    mke2fs hda.img
    sudo mount -o X-mount.mkdir hda.img root
    sudo mkdir -p root/etc/tmpfiles.d
    echo 'd /a/b/c' | sudo tee root/etc/tmpfiles.d/test.conf
    sudo systemd-tmpfiles --root="$PWD/root" --create

Note it prints an error and creates the parents `/a/b` but not the full `/a/b/c`.

Here are some changes that attempt to work around it.  The only change necessary for the problems I encountered is in `path_open_parent_safe`, but I've adjusted `chase_symlinks` elsewhere in tmpfiles so it has consistent behavior.  I'm not very familiar with this code, so if this solution turns out to be unusable, I can close the PR and open an issue instead.

poettering commented:
looks OK, but could you add a unit test for the changes to chase_symlinks() for this? The behaviour on that looks right to me, but I'd really like to see some testing applied to tha, since it's not obvious to see.

dm0- commented:
I've added some tests, which I think pass correctly.  The failed CI checks appear to have had the same errors on other PRs.  I've also added an `unsigned` cast since I saw an error about unsigned comparison in the original change.

martinpitt commented:
This seems to add an extra `/` somewhere? Several `unit-config` tests now fail like this:
```
======================================================================
FAIL: test_unit_alias_enable (__main__.EnableTests)
no sysv: enable unit with an alias
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.s4e55m/build.AVx/systemd/debian/tests/unit-config", line 205, in test_unit_alias_enable
    system_unit_dir + '/test_enable.service')
AssertionError: '/lib/systemd/system/test_enable.service' != '//lib/systemd/system/test_enable.service'
- /lib/systemd/system/test_enable.service
+ //lib/systemd/system/test_enable.service
? +
```

martinpitt commented:
FTR, that "somewhere" is `pkg-config --variable=systemdsystemunitdir systemd`, which apparently gets influenced by this PR to now start with two slashes.

dm0- commented:
@martinpitt I looked at the CI results for the last merged PR, which was merged with the same error: #11827

Looking at the change in the PR, it seems to be the likely cause.

martinpitt commented:
Indeed #11829 has the same error, sorry for the wrong blame. Removing label again.

poettering commented on src/basic/fs-util.c:
```diff
@@ -926,6 +926,7 @@ int chase_symlinks(const char *path, const char *original_root, unsigned flags,
                 if (fstat(child, &st) < 0)
                         return -errno;
                 if ((flags & CHASE_SAFE) &&
+                    (empty_or_root(root) || (unsigned)(todo - buffer) > strlen(root)) &&
```
strlen() returns size_t btw, hence the left-hand cast should be to `(size_t)` too

poettering commented:
looks good, but one typecast nitpick. please fix, will merge then.

poettering commented:
hmm, please always add a comment when you push a new version onto an existing PR. for some reasons github doesn't notify us about that in many cases, so we never notice...
